### ISSUE TITLE
Possibility to disable automatic cache purging/invalidation on resave

### ISSUE THREAD

**Issue by jan-dh** [putyourlightson/craft-blitz#680]
### Support Request

I'm looking into ways to disable the automatic cache purging by Cloudfront on entry resave. I'd like to have more manual control (hourly cache builds & purges for example) to be able to manage Cloudfront costs better. I'd love to run the build process from the CLI or from the Blitz dashboard still. 

Was looking into some of the existing hooks but none of them seems to really help me, closest one I could find was:

```php
Event::on(RefreshCacheService::class, RefreshCacheService::EVENT_BEFORE_ADD_ELEMENT,
	function(RefreshElementEvent $event) {
		Blitz::$plugin->log('Cache purge blocked from resave');
		$event->isValid = false;
	}
);
```

But this just prevents the cache from being saved to the DB as well. 

### Plugin Version

4.18.2

**Comment by bencroker** [comment]
You’ll likely want to use a purge specific event such as [`EVENT_BEFORE_PURGE_CACHE`](https://github.com/putyourlightson/craft-blitz/blob/3e5a4fcecb87b1714b6fe62e7ac791318a708344/src/drivers/purgers/BaseCachePurger.php#L22), which you can use to disable purging of individual URLs. This will still allow purging the entire cache using the `blitz/cache/purge` console command.

```php
use putyourlightson\blitz\Blitz;
use putyourlightson\blitz\drivers\purgers\BaseCachePurger;
use putyourlightson\blitz\events\RefreshCacheEvent;
use yii\base\Event;

Event::on(BaseCachePurger::class, BaseCachePurger::EVENT_BEFORE_PURGE_CACHE,
    function(RefreshCacheEvent $event) {
        $event->isValid = false;
        Blitz::$plugin->log('Cache purging prevented.');
    }
);
```

**Comment by jan-dh** [comment]
We tried that approach but the `blitz/cache/refresh-expired` command doesn't work this way. What we ’d like to do is to be able to run `blitz/cache/refresh-expired` to refresh the cache in Redis & once the cache has been rebuild do a full Cloudfront purge. 

Having the above solution also prevents any entry being submitted to the `blitz_caches` db table.

**Comment by bencroker** [comment]
> We tried that approach but the `blitz/cache/refresh-expired` command doesn't work this way.

What exactly do you mean by “doesn’t work this way”? 

I would think you could create a cron job to run `blitz/cache/refresh-expired` followed by `blitz/cache/purge` and it would all work the way you’d like it to.

**Comment by jan-dh** [comment]
When the event above is in place, saving an entry does not create an entry in the `blitz_caches` db table. Meaning, if there's no entries in the database, the `blitz/cache/refresh-expired` does not work.

**Comment by bencroker** [comment]
> When the event above is in place, saving an entry does not create an entry in the `blitz_caches` db table.

This happens when using my code from https://github.com/putyourlightson/craft-blitz/issues/680#issuecomment-2206280343?

**Comment by jan-dh** [comment]
When using the exact event you posted I get an error in the queue log and the queue task gets stuck and nothing appears in the database.
```
2024-07-03 19:33:15 [queue.ERROR] [TypeError] TypeError: modules\sitemodule\Module::modules\sitemodule\{closure}(): Argument #1 ($event) must be of type modules\sitemodule\RefreshElementEvent, putyourlightson\blitz\events\RefreshCacheEvent given in /var/www/html/modules/sitemodule/Module.php:52
```
When I try to tackle this by using this:

```php
use putyourlightson\blitz\events\RefreshCacheEvent;


Event::on(BaseCachePurger::class, BaseCachePurger::EVENT_BEFORE_PURGE_CACHE,
	function(RefreshCacheEvent $event) {
		$event->isValid = false;
		Blitz::$plugin->log('Cache purging prevented.');
	}
);
```
the event runs on the queue, but nothing appears in the database. 

```
2024-07-03 19:38:22 [queue.INFO] [craft\queue\QueueLogBehavior::afterExec]  [2698701] Refreshing Blitz cache (attempt: 1, pid: 9400) - Done (time: 0.003s) {"memory":27453904} 
```
and the blitz logs gives me
```
[2024-07-03 19:38:22] Cache purging prevented.
```

**Comment by bencroker** [comment]
Ah yes, `RefreshCacheEvent` is indeed the correct event, I’ve updated my code above.

> When the event above is in place, saving an entry does not create an entry in the blitz_caches db table.

Saving an entry does not generally write to the `blitz_caches db` table. Can you please explain what you expect to happen?

**Comment by jan-dh** [comment]
When changing some info to an entry and hit save I would expect it to appear in the `blitz_caches` table to be marked as expired. That way I would assume you can hit `./craft blitz/cache/refresh-expired` to regenerate all the pages that have been updated before doing a cache purge. The scenario I had in mind is:

1. While editors do edits, Blitz keeps track in the DB what pages/elements need to be refreshed. (but preventing automatic Cloudfront purges to keep control over the invalidation costs)
2. Every half hour we generate these elements with the `blitz/cache/refresh-expired` which according to the docs does:

> Refreshing expired cache refreshes all cached pages that have expired, or that contain elements that have expired (applies to elements with future post and expiry dates). Cache duration and expiry dates can be specified in the [config settings](https://putyourlightson.com/plugins/blitz#config-settings) and the [page specific options](https://putyourlightson.com/plugins/blitz#page-specific-options).

3. Once all expired caches are generated (with Redis as target) we would then flush the Cloudfront cache.

By using the `refresh-expired` we want to minimize the server load & only regenerate what actually changed. Otherwise we'd have to do a full cache-build every release. (which we 'd like to set for half an hour/hour).

**Comment by bencroker** [comment]
Yes, I understand your intention. Updating an entry will mark any cached pages (that exist in the `blitz_caches` table) that track that an element (via the `blitz_elementcaches` table) as expired when the refresh mode is set to `Expire the cache...`. What pages and elements are tracked is visible using the [Blitz Diagnostics utility](https://putyourlightson.com/plugins/blitz#diagnostics-utility). Does that make sense now?

**Comment by jan-dh** [comment]
So, implementing the snippet above the following works:
- Updating an entry puts an expired date in the DB
- Refresh expired cache (from the Dashboard)
However, `blitz/cache/refresh-expired` should run the purge function as well, but that is being blocked by the script you mention above.

**Comment by bencroker** [comment]
As I mentioned, if you run `blitz/cache/refresh-expired` (to clear out expired cached pages) followed by `blitz/cache/purge` (to purge the entire cache) then it should all work the way you’d like it to.

**Comment by jan-dh** [comment]
But we only like to purge the expired caches with the `condenseUrl` setting (to prevent unnecessary purges), so running the purge command isn't really a good way out. (sorry if I communicated this unclear above).

Currently trying out things so I can "unblock" the purging if only from a CLI command with specific flag
```php
Event::on(
	BaseCachePurger::class,
	BaseCachePurger::EVENT_BEFORE_PURGE_CACHE,
	function(RefreshCacheEvent $event) {
		Blitz::$plugin->log('CLI command.');
		// Check if this is a console request
		if (Craft::$app->getRequest()->getIsConsoleRequest()) {
			$fullCommand = implode(' ', $_SERVER['argv']);
			if (strpos($fullCommand, '--custom-flag') !== false) {
				Blitz::$plugin->log('Cache purging allowed.');
			}
		}
		// Prevent cache purging for all other requests or CLI commands
		$event->isValid = false;
		Blitz::$plugin->log('Cache purging prevented.');
	}
);
```

**Comment by bencroker** [comment]
Oh, I thought you had previously communicated to me that you wanted to purge the entire cache.

To make this work, I’d suggest you create you own custom console controller action that bypasses the `EVENT_BEFORE_PURGE_CACHE` and purges expired URIs directly. So, in addition to the code in https://github.com/putyourlightson/craft-blitz/issues/680#issuecomment-2206280343, you’d add the following controller and run the `purge-expired` action instead of `blitz/cache/purge`.

```php
use craft\console\Controller;
use putyourlightson\blitz\Blitz;
use putyourlightson\blitz\helpers\SiteUriHelper;
use yii\console\ExitCode;

class PurgeController extends Controller
{
    public function actionPurgeExpired(): int
    {
        $cacheIds = Blitz::$plugin->expireCache->getExpiredCacheIds();
        $siteUris = SiteUriHelper::getCachedSiteUris($cacheIds);
        Blitz::$plugin->cachePurger->purgeUrisWithProgress($siteUris);

        Blitz::$plugin->log('Cache purging completed.');

        return ExitCode::OK;
    }
}
```

**Comment by jan-dh** [comment]
Ok, custom controller implemented and this seems to be moving in the right direction. One big issue we still see is:
- when there are expiry dates in the db, the purges happen only on exactly those
- when there are no expiry dates in the db, everything gets purged

currently looking into something like:
```php
$count = (new \craft\db\Query())
->from('{{%blitz_caches}}')
->where(['not', ['expiryDate' => null]]) 
->andWhere(['!=', 'expiryDate', '']) 
->count();

// If the count is more than 0, it means there are rows matching the criteria
if($count > 0){
	$cacheIds = Blitz::$plugin->expireCache->getExpiredCacheIds();
	$siteUris = SiteUriHelper::getCachedSiteUris($cacheIds);
	Blitz::$plugin->cachePurger->purgeUrisWithProgress($siteUris);

	Blitz::$plugin->log('Cache purging completed.');
} else {
	Blitz::$plugin->log('No expiry dates found, purge skipped.');
}
```
An option to prevent automatic purges would be very much appreciated instead of this hacking around.

**Comment by bencroker** [comment]
You can add that logic as follows.

```php
    public function actionPurgeExpired(): int
    {
        $cacheIds = Blitz::$plugin->expireCache->getExpiredCacheIds();

        if (!empty($cacheIds)) {
            $siteUris = SiteUriHelper::getCachedSiteUris($cacheIds);
            Blitz::$plugin->cachePurger->purgeUrisWithProgress($siteUris);
            Blitz::$plugin->log('Cache purging completed.');
        }

        return ExitCode::OK;
    }
```

The `EVENT_BEFORE_PURGE_CACHE` exists specifically to allow you to prevent purges. Since you have a very edge-case need, however, a custom workaround is required to be able to determine exactly when purging occurs.

**Comment by jan-dh** [comment]
Thanks, testing this. This setup could be expanded upon by hooking into this controller and after the purge use blitz cache igniter to warm the purged urls I assume? Is the `$siteUris` something I could pass to `php craft cache-igniter/warm/urls`?

**Comment by bencroker** [comment]
Absolutely! Here’s how you can warm the expired URLs directly after purging. 

```php
use putyourlightson\blitz\helpers\SiteUriHelper;
use putyourlightson\cacheigniter\CacheIgniter;

$urls = SiteUriHelper::getUrlsFromSiteUris($siteUris);
CacheIgniter::$plugin->warmer->warmUrlsWithProgress($urls);
```

I suggest you put all of the logic you want to execute in this one controller action, so you can fully control the process. This way you don’t need to run `blitz/cache/refresh-expired`. Here’s what the entire controller might look like.

```php
use craft\console\Controller;
use putyourlightson\blitz\Blitz;
use putyourlightson\blitz\helpers\SiteUriHelper;
use putyourlightson\cacheigniter\CacheIgniter;
use yii\console\ExitCode;

class CustomController extends Controller
{
    public function actionRefreshExpired(): int
    {
        $cacheIds = Blitz::$plugin->expireCache->getExpiredCacheIds();
        $siteUris = SiteUriHelper::getCachedSiteUris($cacheIds);

        if (empty($siteUris)) {
            return ExitCode::OK;
        }
        
        Blitz::$plugin->cacheStorage->deleteUris($siteUris);
        Blitz::$plugin->log('Cache cleared.');
        
        Blitz::$plugin->cachePurger->purgeUrisWithProgress($siteUris);
        Blitz::$plugin->log('Cache purged.');

        $urls = SiteUriHelper::getUrlsFromSiteUris($siteUris);
        CacheIgniter::$plugin->warmer->warmUrlsWithProgress($urls);
        Blitz::$plugin->log('Cache warmed.');

        return ExitCode::OK;
    }
}
```

**Comment by jan-dh** [comment]
This setup is currently not blocking the SEOmatic integration:
```
'putyourlightson\blitz\drivers\integrations\SeomaticIntegration',
```
If I'd disable the integration, will SEOmatic caches still get cleared with the manual clear?

**Comment by bencroker** [comment]
It should still block individual URLs from being purged, even if initiated from the SEOmatic integration. The only thing that will go through is refreshing the entire cache, which happens whenever SEOmatic settings are changed. 

If you prefer to prevent this entirely, you can disable the SEOmatic integration by uncommenting the `integrations` config setting and remove the `SeomaticIntegration` class. SEOmatic will continue to work as expected but Blitz will be unaware of any changes made to it that may affect cached pages.
https://github.com/putyourlightson/craft-blitz/blob/c3f4495e1b96b07950d057e45fe00d0152f71552/src/config.php#L226-L230

**Comment by jan-dh** [comment]
I updated some info on an SEOmatic field linked to an entry, which then triggered a `Generating sitemap` task:
![image (8)](https://github.com/putyourlightson/craft-blitz/assets/5801781/d37a346b-b973-4bdc-a472-5822391984c2)
which then did a full cache invalidation.

Disabling the integration will also not clear caches for pages with my cronjob if only SEOmatic info updates.

**Comment by bencroker** [comment]
That’s correct, with the integration disabled, Blitz is unaware of any changes made to SEOmatic.

**Comment by bencroker** [comment]
It’s possible to disable purging all cache using the [`EVENT_BEFORE_PURGE_ALL_CACHE`](https://github.com/putyourlightson/craft-blitz/blob/c3f4495e1b96b07950d057e45fe00d0152f71552/src/drivers/purgers/BaseCachePurger.php#L32) in addition to `EVENT_BEFORE_PURGE_CACHE` as above.

**Comment by jan-dh** [comment]
Tried `EVENT_BEFORE_PURGE_CACHE` but it's not blocking, do you need to combine the 2? (as you say - in addition to)

**Comment by bencroker** [comment]
Yes, you can block both purging of individual URLs using `EVENT_BEFORE_PURGE_CACHE` and the entire cache using `EVENT_BEFORE_PURGE_ALL_CACHE`, if you wish.

**Comment by jan-dh** [comment]
Hmm, still some events coming through, what I test is:
- updating an SEO field on an entry
- reordering elements in a structure
- update SEOmatic content settings
I assume that the first 2 I can trigger by updating that entry on another field. The last one, not sure with the SEOmatic integration disabled how I would be able to trigger that?

**Comment by bencroker** [comment]
What exactly are you trying to achieve now?

**Comment by jan-dh** [comment]
The same behaviour we have for the normal content resaves etc. I'd like to block SEOmatic from triggering invalidations  to Cloudfront (when doing one of the actions I'm testing above). I only want invalidations to run through the custom controller we run with a cron job.

```php
class CustomConsoleController extends Controller
{
    public function actionPurgeExpired()
    {
		Blitz::$plugin->log('Action Purge expired triggered.');
		$cacheIds = Blitz::$plugin->expireCache->getExpiredCacheIds();
		$siteUris = SiteUriHelper::getCachedSiteUris($cacheIds);

		if (empty($cacheIds)) {
			Blitz::$plugin->log('No expiry dates found, purge skipped.');
			return ExitCode::OK;
		}

		Blitz::$plugin->log('CLI Command ran.');
		$command = 'blitz/cache/refresh-expired';
		Craft::$app->runAction($command);

		Blitz::$plugin->cachePurger->purgeUrisWithProgress($siteUris);
		Blitz::$plugin->log('Cache purging completed.');

		return ExitCode::OK;
    }
}
```

**Comment by bencroker** [comment]
Combining the `actionRefreshExpired()` method I provided in https://github.com/putyourlightson/craft-blitz/issues/680#issuecomment-2210466313, as well as blocking automatic purging using the following events, should work.

```php
use putyourlightson\blitz\Blitz;
use putyourlightson\blitz\drivers\purgers\BaseCachePurger;
use putyourlightson\blitz\events\RefreshCacheEvent;
use yii\base\Event;

Event::on(BaseCachePurger::class, BaseCachePurger::EVENT_BEFORE_PURGE_CACHE,
    function(RefreshCacheEvent $event) {
        $event->isValid = false;
        Blitz::$plugin->log('Cache purging prevented.');
    }
);

Event::on(BaseCachePurger::class, BaseCachePurger::EVENT_BEFORE_PURGE_ALL_CACHE,
    function(RefreshCacheEvent $event) {
        $event->isValid = false;
        Blitz::$plugin->log('Cache purging prevented.');
    }
);
```

**Comment by jan-dh** [comment]
Yep, this is what I've tried but I still see an invalidation request coming in to Cloudfront, will test further to see what exactly causes it.

**Comment by jan-dh** [comment]
Alright, seems it's ok, my bad.

**Comment by bencroker** [comment]
So is everything now working the way you want it to?

**Comment by jan-dh** [comment]
Almost, we had some issues running the generate cache in combination with the `token` param.  In general, it's a bit of a search to see what params should be whitelisted in Cloudfront to not be cached to make sure the live preview etc. keeps working. Currently the setup was to generate a unique cache for urls with
```
x-craft-preview
x-craft-live-preview
token
```
Ideally these would be exempted from the cache but that exposes the site from a lot of spam with these URL params that go around the cache I guess.

**Comment by bencroker** [comment]
Not sure what you mean, but any live preview requests or requests with tokens should be excluded from caching by defualt by Blitz.

**Comment by bencroker** [comment]
Closing this issue as the initial question appears to be resolved. If any related issues come up, please open a new issue with specifics.
